<a href="https://colab.research.google.com/github/AARomanow/wiki/blob/main/Ali_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
! wget https://sourceforge.net/projects/bowtie-bio/files/bowtie2/2.4.1/bowtie2-2.4.1-linux-x86_64.zip/download -O bowtie2-2.4.1.zip

--2023-08-14 07:04:30--  https://sourceforge.net/projects/bowtie-bio/files/bowtie2/2.4.1/bowtie2-2.4.1-linux-x86_64.zip/download
Resolving sourceforge.net (sourceforge.net)... 104.18.11.128, 104.18.10.128, 2606:4700::6812:a80, ...
Connecting to sourceforge.net (sourceforge.net)|104.18.11.128|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://downloads.sourceforge.net/project/bowtie-bio/bowtie2/2.4.1/bowtie2-2.4.1-linux-x86_64.zip?ts=gAAAAABk2dH_aazLY3moAzoGwk32cwYG5AGCYt0fD3sejjYdRC6Py1ZT0lxc4WzdUha26A7vIkeEFMhbW4lElRrupj58brofSA%3D%3D&use_mirror=versaweb&r= [following]
--2023-08-14 07:04:31--  https://downloads.sourceforge.net/project/bowtie-bio/bowtie2/2.4.1/bowtie2-2.4.1-linux-x86_64.zip?ts=gAAAAABk2dH_aazLY3moAzoGwk32cwYG5AGCYt0fD3sejjYdRC6Py1ZT0lxc4WzdUha26A7vIkeEFMhbW4lElRrupj58brofSA%3D%3D&use_mirror=versaweb&r=
Resolving downloads.sourceforge.net (downloads.sourceforge.net)... 204.68.111.105
Connecting to downloads.sourceforge.net (downloads.

In [ ]:
!unzip bowtie2-2.4.1.zip

Archive:  bowtie2-2.4.1.zip
   creating: bowtie2-2.4.1-linux-x86_64/
  inflating: bowtie2-2.4.1-linux-x86_64/bowtie2-inspect-l  
   creating: bowtie2-2.4.1-linux-x86_64/example/
   creating: bowtie2-2.4.1-linux-x86_64/example/reference/
  inflating: bowtie2-2.4.1-linux-x86_64/example/reference/lambda_virus.fa  
   creating: bowtie2-2.4.1-linux-x86_64/example/reads/
  inflating: bowtie2-2.4.1-linux-x86_64/example/reads/simulate.pl  
  inflating: bowtie2-2.4.1-linux-x86_64/example/reads/longreads.fq  
  inflating: bowtie2-2.4.1-linux-x86_64/example/reads/combined_reads.bam  
  inflating: bowtie2-2.4.1-linux-x86_64/example/reads/reads_2.fq  
  inflating: bowtie2-2.4.1-linux-x86_64/example/reads/reads_1.fq  
   creating: bowtie2-2.4.1-linux-x86_64/example/index/
  inflating: bowtie2-2.4.1-linux-x86_64/example/index/lambda_virus.rev.2.bt2  
  inflating: bowtie2-2.4.1-linux-x86_64/example/index/lambda_virus.1.bt2  
  inflating: bowtie2-2.4.1-linux-x86_64/example/index/lambda_virus.3.bt2  
  

In [ ]:
import os
os.environ['PATH'] += ":/content/bowtie2-2.4.1-linux-x86_64"


In [ ]:
!ls /content/


bowtie2-2.4.1-linux-x86_64  bowtie2_index.tgz  genome.fa	  sample_data
bowtie2-2.4.1.zip	    e_coli.gtf	       reads_7.fastq.tgz


In [ ]:
import os
os.environ['PATH'] += ":/content/bowtie2-2.4.1-linux-x86_64"


In [ ]:
!bowtie2 --version


/content/bowtie2-2.4.1-linux-x86_64/bowtie2-align-s version 2.4.1
64-bit
Built on 
Fri Feb 28 22:21:25 UTC 2020
Compiler: gcc version 7.3.1 20180303 (Red Hat 7.3.1-5) (GCC) 
Options: -O3 -msse2 -funroll-loops -g3 -g -O2 -fvisibility=hidden -I/hbb_exe_gc_hardened/include -ffunction-sections -fdata-sections -fstack-protector -D_FORTIFY_SOURCE=2 -fPIE -DPOPCNT_CAPABILITY -DWITH_TBB -std=c++11 -DNO_SPINLOCK -DWITH_QUEUELOCK=1
Sizeof {int, long, long long, void*, size_t, off_t}: {4, 8, 8, 8, 8, 8}


In [ ]:
!wget https://sourceforge.net/projects/subread/files/subread-2.0.1/subread-2.0.1-Linux-x86_64.tar.gz/download -O subread-2.0.1.tar.gz


--2023-08-14 07:08:34--  https://sourceforge.net/projects/subread/files/subread-2.0.1/subread-2.0.1-Linux-x86_64.tar.gz/download
Resolving sourceforge.net (sourceforge.net)... 104.18.11.128, 104.18.10.128, 2606:4700::6812:a80, ...
Connecting to sourceforge.net (sourceforge.net)|104.18.11.128|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://downloads.sourceforge.net/project/subread/subread-2.0.1/subread-2.0.1-Linux-x86_64.tar.gz?ts=gAAAAABk2dLyk9muVby4mYgbrXH3gmCrxusvlMety7pVVys1LL4IvksTp1Demr1RjI55rriiVqn0xqSdsaavVW7flqe1cWJyuQ%3D%3D&use_mirror=cfhcable&r= [following]
--2023-08-14 07:08:34--  https://downloads.sourceforge.net/project/subread/subread-2.0.1/subread-2.0.1-Linux-x86_64.tar.gz?ts=gAAAAABk2dLyk9muVby4mYgbrXH3gmCrxusvlMety7pVVys1LL4IvksTp1Demr1RjI55rriiVqn0xqSdsaavVW7flqe1cWJyuQ%3D%3D&use_mirror=cfhcable&r=
Resolving downloads.sourceforge.net (downloads.sourceforge.net)... 204.68.111.105
Connecting to downloads.sourceforge.net (downloads.

In [ ]:
!tar -xf subread-2.0.1.tar.gz


In [ ]:
os.environ['PATH'] += ":/content/subread-2.0.1-Linux-x86_64/bin/"


In [ ]:
!bowtie2 --version


/content/bowtie2-2.4.1-linux-x86_64/bowtie2-align-s version 2.4.1
64-bit
Built on 
Fri Feb 28 22:21:25 UTC 2020
Compiler: gcc version 7.3.1 20180303 (Red Hat 7.3.1-5) (GCC) 
Options: -O3 -msse2 -funroll-loops -g3 -g -O2 -fvisibility=hidden -I/hbb_exe_gc_hardened/include -ffunction-sections -fdata-sections -fstack-protector -D_FORTIFY_SOURCE=2 -fPIE -DPOPCNT_CAPABILITY -DWITH_TBB -std=c++11 -DNO_SPINLOCK -DWITH_QUEUELOCK=1
Sizeof {int, long, long long, void*, size_t, off_t}: {4, 8, 8, 8, 8, 8}


In [ ]:
!featureCounts -v




featureCounts v2.0.1



# Question 1

Understanding the FASTQ Format

FASTQ format is a text-based format for representing either nucleotide sequences or protein sequences, together with their quality scores. One entry in a FASTQ file consists of four lines:

- Line 1: Begins with a @ character and is followed by a sequence identifier and an optional description.
- Line 2: The raw sequence letters.
- Line 3: Begins with a + character and is optionally followed by the sequence identifier and an optional description (this line can be just a +).
- Line 4: Encodes the quality values for the sequence in Line 2, and must contain the same number of symbols as the length of the sequence. A quality score of the i-th base is the ASCII value of the i-th character in the fourth line minus 33.
Interpreting a FASTQ Entry

In [ ]:
!tar -xzf reads_7.fastq.tgz

In [ ]:
# First, let's unpack the reads file
!tar -xzf reads_7.fastq.tgz

# Extract one read
with open("reads_008_R1.fastq", "r") as file:
    for _ in range(4):
        print(file.readline().strip())


@Chromosome_0_0/1
ATAATGCGCAGGAACATTACCAACGGATAGACAGTGGCGTAAGAGAGCGCCGCCGCACCGCTGGTTGGATGAAGATTATTAGCAAACGCCAGCGCCGGAGGATCGGTCATGGAACCTGCCAGCATCCCGCACATGGTCAGGTAATTCATTTTGGCTAACATCCGCGCCAGAATGCCAACAGTAATCAGCGGAACGGCGGTGATCAGGGCACCATAACCAATCCAGCTTAGCCCTTCGCCATTGACCAGGGTATTCACAAAATTCCCACCAGATTTCCGACCAACGACCGAGTGGAACAGTA
+
CCCCCGGGGGGGGGGGGGFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGEGGGGGGGGGGGGGGGGGGDEGGGGGGFGGGGGGGGEGFEGGGFGGGGGCGGGFCGFGGGGGGFGGGFGFGGGGFGGGF=GGGGGGGGGCGGGGCFFGGEGFGGCGFGG9GGGGGGDFFGGGFGGGGG@:GEGGFGEGGGGGGEFCGFFGGG>EE@EG<GCGG5GGGGC2E<FCGFGGEG?9EG747FCE7FCF<GGDGG3*G/FFG::+EE/3G/*+777:))FFF7<=F5A11*7))F.*03C1**


1. Read Identifier:
The sequence begins with @Chromosome_0_0/1. This denotes the unique name or identifier for the read. Here, "Chromosome_0_0" could be a naming convention employed by the sequencing platform or a representation of the source. The "/1" is indicative of the possibility that this could be the first read in paired-end sequencing.


2. Nucleotide Sequence:
Following the identifier is the nucleotide sequence:

ATAATGCGCAGGAACATTACCAACGGATAGACAGTGGCGTAAGAGAGCGCCGCCGCACCGCTGGTTGGATGAAGATTATTAGCAAACGCCAGCGCCGGAGGATCGGTCATGGAACCTGCCAGCATCCCGCACATGGTCAGGTAATTCATTTTGGCTAACATCCGCGCCAGAATGCCAACAGTAATCAGCGGAACGGCGGTGATCAGGGCACCATAACCAATCCAGCTTAGCCCTTCGCCATTGACCAGGGTATTCACAAAATTCCCACCAGATTTCCGACCAACGACCGAGTGGAACAGTA


3. Quality Score Delimiter:
The character + acts as a separator between the nucleotide sequence and its associated quality scores. It simply demarcates the beginning of the quality scores.

4. Quality Scores:
The subsequent string of characters encodes the quality of the nucleotide sequence. Each symbol signifies a Phred quality score for the corresponding nucleotide in the sequence. These scores are ASCII encoded.

For example, the character C gives us a Phred quality score of 34 (ASCII value of C - 33). A Phred score of 34 indicates that the likelihood of an error in base calling is quite low.

It's worth noting that the higher the Phred score, the lower the error probability:

A Phred score of 20: Implies a 1% error probability.
A Phred score of 30: Indicates a 0.1% error probability.
A Phred score of 40: Corresponds to a 0.01% error probability.
In our given read, the initial bases possess high-quality scores (represented by numerous G's). This means these bases were read with great confidence. Towards the end, the quality scores see a reduction, suggesting a decline in base call confidence.


Conclusion:
A FASTQ entry provides both the nucleotide sequence read by a sequencer and a measure of confidence (quality scores) for each base in that sequence. Analyzing these quality scores is pivotal for bioinformatics workflows as it aids in filtering and refining the data, ensuring its reliability for subsequent analyses.


# Task 2

In [ ]:
!tar -xzf bowtie2_index.tgz


In [ ]:
!bowtie2-build genome.fa ecoli



Settings:
  Output files: "ecoli.*.bt2"
  Line rate: 6 (line is 64 bytes)
  Lines per side: 1 (side is 64 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: unpacked
  Max bucket size: default
  Max bucket size, sqrt multiplier: default
  Max bucket size, len divisor: 4
  Difference-cover sample period: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: disabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  genome.fa
Building a SMALL index
Reading reference sizes
  Time reading reference sizes: 00:00:00
Calculating joined length
Writing header
Reserving space for joined string
Joining reference sequences
  Time to join reference sequences: 00:00:00
bmax according to bmaxDivN setting: 1160413
Using parameters --bmax 870310 --dcv 1024
  Doing ahead-of-time memory usage test
  Passed!  Constructing with these parameters: --bmax 870310 --dcv 1024
Constructing suffix-array element gene

In [ ]:
!bowtie2 -x ecoli -U reads_008_R1.fastq -S output.sam


10000 reads; of these:
  10000 (100.00%) were unpaired; of these:
    0 (0.00%) aligned 0 times
    9758 (97.58%) aligned exactly 1 time
    242 (2.42%) aligned >1 times
100.00% overall alignment rate


In [ ]:
!grep -v '^@' output.sam | head -1


Chromosome_0_0/1	0	Chromosome	3864655	42	301M	*	0	0	ATAATGCGCAGGAACATTACCAACGGATAGACAGTGGCGTAAGAGAGCGCCGCCGCACCGCTGGTTGGATGAAGATTATTAGCAAACGCCAGCGCCGGAGGATCGGTCATGGAACCTGCCAGCATCCCGCACATGGTCAGGTAATTCATTTTGGCTAACATCCGCGCCAGAATGCCAACAGTAATCAGCGGAACGGCGGTGATCAGGGCACCATAACCAATCCAGCTTAGCCCTTCGCCATTGACCAGGGTATTCACAAAATTCCCACCAGATTTCCGACCAACGACCGAGTGGAACAGTA	CCCCCGGGGGGGGGGGGGFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGEGGGGGGGGGGGGGGGGGGDEGGGGGGFGGGGGGGGEGFEGGGFGGGGGCGGGFCGFGGGGGGFGGGFGFGGGGFGGGF=GGGGGGGGGCGGGGCFFGGEGFGGCGFGG9GGGGGGDFFGGGFGGGGG@:GEGGFGEGGGGGGEFCGFFGGG>EE@EG<GCGG5GGGGC2E<FCGFGGEG?9EG747FCE7FCF<GGDGG3*G/FFG::+EE/3G/*+777:))FFF7<=F5A11*7))F.*03C1**	AS:i:-9	XN:i:0	XM:i:4	XO:i:0	XG:i:0	NM:i:4	MD:Z:262C13A14A7C1	YT:Z:UU


**Task 2 Analysis: Dissecting a Read in SAM Format**

Given SAM Entry:

```
Chromosome_0_0/1	0	Chromosome	3864655	42	301M	*	0	0	ATAATGCGCAGGAACATTACCAACGGATAGACAGTGGCGTAAGAGAGCGCCGCCGCACCGCTGGTTGGATGAAGATTATTAGCAAACGCCAGCGCCGGAGGATCGGTCATGGAACCTGCCAGCATCCCGCACATGGTCAGGTAATTCATTTTGGCTAACATCCGCGCCAGAATGCCAACAGTAATCAGCGGAACGGCGGTGATCAGGGCACCATAACCAATCCAGCTTAGCCCTTCGCCATTGACCAGGGTATTCACAAAATTCCCACCAGATTTCCGACCAACGACCGAGTGGAACAGTA	CCCCCGGGGGGGGGGGGGFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGEGGGGGGGGGGGGGGGGGGDEGGGGGGFGGGGGGGGEGFEGGGFGGGGGCGGGFCGFGGGGGGFGGGFGFGGGGFGGGF=GGGGGGGGGCGGGGCFFGGEGFGGCGFGG9GGGGGGDFFGGGFGGGGG@:GEGGFGEGGGGGGEFCGFFGGG>EE@EG<GCGG5GGGGC2E<FCGFGGEG?9EG747FCE7FCF<GGDGG3*G/FFG::+EE/3G/*+777:))FFF7<=F5A11*7))F.*03C1**	AS:i:-9	XN:i:0	XM:i:4	XO:i:0	XG:i:0	NM:i:4	MD:Z:262C13A14A7C1	YT:Z:UU
```

Let's break down the provided SAM format entry:

1. **QNAME: Chromosome_0_0/1**  
   This represents the query (read) name. In this case, the read is named `Chromosome_0_0/1`.

2. **FLAG: 0**  
   The flag field provides a bitwise representation of different properties of the read. A value of `0` generally indicates that the read is mapped in a forward orientation.

3. **RNAME: Chromosome**  
   This is the reference sequence name where the read has been mapped. The read mapped onto a reference named "Chromosome".

4. **POS: 3864655**  
   It denotes the 1-based leftmost position/coordinate of the clipped alignment.

5. **MAPQ: 42**  
   This is the mapping quality. It represents the Phred-scaled probability that the mapping position of this read is incorrect. The higher the score, the more confidence there is in the mapping.

6. **CIGAR: 301M**  
   This represents the CIGAR string. The `301M` implies that all 301 bases in the read match to the reference genome.

7. **RNEXT: ***  
   The reference name of the mate/next read. `*` indicates that the information is not available.

8. **PNEXT: 0**  
   The position of the mate/next read. A value of `0` indicates it's not available or not applicable.

9. **TLEN: 0**  
   This is the observed template length. `0` can mean the mate pair information is not available or the read is single-ended.

10. **SEQ: ATAATGCGCA...**  
    This is the read's sequence.

11. **QUAL: CCC...**  
    It represents the Phred quality score of each base in the read. A high score (e.g., close to `Q` in ASCII) indicates a high confidence in the base call, and a lower score (e.g., close to `!`) indicates a lower confidence.

12. **TAGs**  
   - **AS:i:-9** - Alignment score.
   - **XN:i:0** - Total number of ambiguous positions in alignment.
   - **XM:i:4** - Number of mismatches in the alignment.
   - **XO:i:0** - Number of gap openings.
   - **XG:i:0** - Total number of gaps (gap extends + gap opens).
   - **NM:i:4** - Total number of mismatches and gap openings.
   - **MD:Z:262C13A14A7C1** - String for mismatching positions.
   - **YT:Z:UU** - SAM tag indicating the type of read. Typically used by tools like Tophat.

**Analysis Conclusion:**  
The read "Chromosome_0_0/1" aligns to the reference "Chromosome" starting at position 3864655. The entire length of the read matches the reference genome with no insertions or deletions. There are four mismatches in this alignment, as indicated by the `XM` and `MD` tags. The mapping quality of 42 suggests a high confidence in this alignment's accuracy.



**Task 3 Analysis: Discrepancy in Read Count by featureCounts**

When working with RNA-seq data and tools like featureCounts to assign reads to genomic features, it's common to observe that not all reads are counted. Several reasons might explain this discrepancy:

1. **Ambiguous Mapping**: Some reads might align to multiple locations in the genome with similar alignment scores. If the tool can't confidently determine the unique origin of a read, it might discard or not count it.

2. **Mapping Outside of Features**: featureCounts assigns reads to known genomic features, like exons or genes. If a read maps to an intergenic region or an intronic region (and you're counting exonic reads), it won't be counted.

3. **Multimapping Reads**: If a read maps to multiple genes, featureCounts might discard it, depending on the settings. This helps avoid ambiguous assignment of reads to genes.

4. **Quality Thresholds**: Reads with low mapping quality might be discarded. Poor quality alignments are less trustworthy, and removing them can help increase the confidence in the results.

5. **Overlapping Genes**: Sometimes, two genes might overlap in a genome. Reads that map to these overlapping regions can be hard to assign confidently to one gene or another. Depending on settings, such reads might be discarded or counted for both genes.

6. **Strand-specific Protocol**: If the RNA-seq was conducted using a strand-specific protocol and the settings of featureCounts are set to consider strand information, then reads mapping to the wrong strand will not be counted.

7. **Improper Paring in Paired-end Sequencing**: In paired-end sequencing, if one of the read pairs maps well but its mate does not, the pair might be discarded depending on the settings.

8. **Genome Annotation Discrepancies**: The annotation file (like GTF or GFF) used with featureCounts determines which features are recognized. If the annotation is outdated or incomplete, reads mapping to unannotated features won't be counted.

**Conclusion**:  
While featureCounts is a powerful tool for assigning reads to genomic features, several reasons can lead to some reads not being counted. It's essential to understand these reasons, as they impact the interpretation of RNA-seq results. Evaluating the summary output of featureCounts can provide insights into the number of reads discarded and why, helping to address these discrepancies.



In [ ]:
!# Extract top 3 genes with highest mapped reads
!awk 'NR>2 {print $1, $NF}' counts.txt | sort -k2,2nr | head -3 > top3_genes.txt


awk: cannot open counts.txt (No such file or directory)


In [ ]:
!ls

bowtie2-2.4.1-linux-x86_64  ecoli.3.bt2       genome.fa
bowtie2-2.4.1.zip	    e_coli.4.bt2      output.sam
bowtie2_index.tgz	    ecoli.4.bt2       reads_008_R1.fastq
e_coli.1.bt2		    e_coli.gtf	      reads_7.fastq.tgz
ecoli.1.bt2		    e_coli.rev.1.bt2  sample_data
e_coli.2.bt2		    ecoli.rev.1.bt2   subread-2.0.1-Linux-x86_64
ecoli.2.bt2		    e_coli.rev.2.bt2  subread-2.0.1.tar.gz
e_coli.3.bt2		    ecoli.rev.2.bt2   top3_genes.txt


In [ ]:
!bowtie2 -x e_coli -U reads_008_R1.fastq -S output.sam


10000 reads; of these:
  10000 (100.00%) were unpaired; of these:
    0 (0.00%) aligned 0 times
    9758 (97.58%) aligned exactly 1 time
    242 (2.42%) aligned >1 times
100.00% overall alignment rate


In [ ]:
!subread-2.0.1-Linux-x86_64/bin/featureCounts -a e_coli.gtf -o counts.txt output.sam



        ==========     _____ _    _ ____  _____  ______          _____  
        =====         / ____| |  | |  _ \|  __ \|  ____|   /\   |  __ \ 
          =====      | (___ | |  | | |_) | |__) | |__     /  \  | |  | |
            ====      \___ \| |  | |  _ <|  _  /|  __|   / /\ \ | |  | |
              ====    ____) | |__| | |_) | | \ \| |____ / ____ \| |__| |
        ==========   |_____/ \____/|____/|_|  \_\______/_/    \_\_____/
	  v2.0.1

//========================== featureCounts setting ===========================\\
||                                                                            ||
||             Input files : 1 SAM file                                       ||
||                           o output.sam                                     ||
||                                                                            ||
||             Output file : counts.txt                                       ||
||                 Summary : counts.txt.summary                  

In [ ]:
!awk 'NR>2 {print $1, $NF}' counts.txt | sort -k2,2nr | head -3 > top3_genes.txt


In [ ]:
!cat top3_genes.txt


b0497 16
b1978 15
b1378 14



### Task 4 [3 marks]

**Objective**: To identify the three genes with the highest number of mapped reads using bowtie2, featureCounts, and subsequent result processing.

**Methodology**:
1. **Read Mapping**: Used `bowtie2` for aligning reads to the reference genome, resulting in a SAM file (`output.sam`).
2. **Gene-based Read Counting**: Used `featureCounts` to count the number of reads mapped to each gene based on the provided GTF file (`e_coli.gtf`). The results were stored in `counts.txt`.
3. **Result Processing**: Parsed the `counts.txt` file to extract and sort genes based on read counts.

**Results**:
From the processed data, the three genes with the highest mapped read counts are:
1. **Gene b0497** - **16 reads**
2. **Gene b1978** - **15 reads**
3. **Gene b1378** - **14 reads**

